javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -1387,23 +1265,158 @@ public String toString() {
+                    newArray[newArray.length - 1] = array;
+                    array = newArray;
+
+                    depthShift += BRANCHING_BASE;
+                    offset = treeSize(branchingFactor() - 1, depthShift);
+                }
+
+                offset -= 1;
+                array = recursiveSet(array, offset, depthShift, leading, 0);
+            }
+
+            return new VectorTree<T>(array, offset, newSize, depthShift);
+        }
+
+        VectorTree<T> appendLeaf(Object trailing) {
+            final int newSize = length() + 1;
+            int depthShift = this.depthShift;
+            Object[] array = asArray(trailing);
+            if (length() > 0) {
+                array = this.array;
+                if ((offset + newSize) > treeSize(branchingFactor(), depthShift)) {
+                    array = asArray(array);
+                    depthShift += BRANCHING_BASE;
+                }
+
+                array = recursiveSet(array, offset + length(), depthShift, trailing, 0);
+            }
+
+            return new VectorTree<T>(array, offset, newSize, depthShift);
+        }
+
+        VectorTree<T> update(int index, T element) {
+            final Object[] root = recursiveSet(array, offset + index, depthShift, element, 0);
+            return new VectorTree<T>(root, offset, length(), depthShift);
+        }
+
+        /** Since the depth of the tree is always small, a non-tailrecursive call is actually faster than pre-calculating the path (is probably unrolled) */
+        static Object[] recursiveSet(Object[] array, int index, int depthShift, Object element, int endShift) {
+            final int childIndex = digit(index, depthShift);
+            if (depthShift <= endShift) {
+                return copyUpdate(array, childIndex, element);
             } else {
-                return elements.zipWithIndex().flatMap(t -> apply(elements.drop(t._2 + 1), (k - 1))
-                        .map((Vector<T> c) -> c.prepend(t._1))
-                );
+                Object[] childArray = getOrDefault(array, childIndex, emptyArray());
+                childArray = recursiveSet(childArray, index, depthShift - BRANCHING_BASE, element, endShift);
+                return copyUpdate(array, childIndex, childArray);
             }
         }
+
+        VectorTree<T> drop(int n) {
+            final int index = offset + n;
+
+            int shiftValue = depthShift;
+            int previousOffset = firstDigit(index, shiftValue);
+            final Object[] root = copyDrop(array, previousOffset);
+            for (Object[] array = root; (array != null) && (shiftValue >= BRANCHING_BASE); ) {
+                shiftValue -= BRANCHING_BASE;
+                final int offset = digit(index, shiftValue);
+
+                final Object[] newNode = copyDrop(getAt(array, previousOffset), offset);
+                array[previousOffset] = newNode;
+                array = newNode;
+
```

Comment:
... you did point out an asymmetry though :)
`drop` and `take` should have had a similar structure (one was iterative, the other recursive).
By changing `take` to be iterative (i.e. the equivalent of a tail recursive call) it became more unified and almost twice as fast for big values :).
Thanks.